File: LeetCode/3. Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int res = 0;
        Map<Character, Integer> map = new HashMap<>(){
            @Override
            public Integer get(Object key) {
                if(!containsKey(key))
                    return 0;
                return super.get(key);
            }
        }; //char : index

        int start = 0;
        for (int end = 0; end < s.length(); end++) {
            char c = s.charAt(end);
            if (map.containsKey(c)) {
                start = Math.max(start, map.get(c) + 1);
            }
            map.put(c, end);
            res = Math.max(res, end - start + 1);
        }
        return res;
    }
} //TC: O(n), SC: O(n) as hash map is used, where n is the length of the string

File: LeetCode/3. Longest Substring Without Repeating Characters.kt
 class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        val map = HashMap<Char, Int>() //char:latest index
        var head = 0
        var tail = 0

        var res = 0

        while (tail < s.length) { //loop until the tail index is over length of the string
            if (s[tail] in map) { //if current char is repeated
                head = Math.max(head, map[s[tail]]!!+1) //update the head index to the last repeated char index+1 if it is larger
            }
            map[s[tail]] = tail //update the char index to the latest one
            ++tail //increment for the tail index
            res = Math.max(res, tail-head) //update the result if current length is larger
        }
        return res
    }
}
